### PR TITLE
[INTEL MKL] Add primitive cache for mkl concat

### PR DESCRIPTION
**Pull request by pallavigopal**
Add support for primitive reuse in mkl concat with all the inputs as TF tensors.

Files modified
tensorflow/core/kernels/mkl_concat_op.cc

Please review

### CODE REVIEW

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 1]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+  MklConcatFwdPrimitiveFactory() {}
+  ~MklConcatFwdPrimitiveFactory() {}
+
+  static MklConcatFwdPrimitiveFactory& GetInstance() {
+    static MklConcatFwdPrimitiveFactory instance_;
+    return instance_;
+  }
+
+  static string CreateKey(const MklConcatFwdParams& concat_fwd_dims) {
+    string prefix = "Concat_fwd_";
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_concat_op.cc]: Other files seem to use all lowercase letters for keys. Let's make this consistent.
```suggestion
    string prefix = "concat_fwd_";
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR and sorry for my delay! I have minor formatting comments.

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 2]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+        mkl_common_format(mkl_common_format) {
+    for (int k = 0; k < num_inputs; k++) {
+      src_dims.push_back(src_dims_pt[k]);
+    }
+  }
+};
+
+// TODO(intel-tf): The template type "T" is currently used to match the
+// templatized class MklPrimitiveFactory (tensorflow/core/util/mkl_util.h).
+// In future, with the removal of "T" from MklPrimitiveFactory, this class
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 3]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+  }
+
+ private:
+  // Primitive reuse context for concat Fwd op
+  struct ConcatFwdContext {
+    std::vector<mkldnn::memory::primitive_desc> src_pd;
+    std::vector<std::shared_ptr<mkldnn::memory::primitive_desc>> src_pd_shdptr;
+    std::shared_ptr<mkldnn::memory::primitive_desc> dst_pd;
+
+    // MKLDNN memory
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 4]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+    std::vector<mkldnn::primitive::at> data_mem;
+    std::vector<std::shared_ptr<mkldnn::memory>> data_mem_shdptr;
+    std::shared_ptr<mkldnn::memory> dst_mem;
+
+    // memory desc
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 5]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+    std::vector<std::shared_ptr<mkldnn::memory::desc>> src_md;
+    std::shared_ptr<mkldnn::memory::desc> dst_md;
+
+    //  concat primitive
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 6]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+  memory::format mkl_common_format;
+
+  MklConcatFwdParams(std::vector<memory::dims>& src_dims_pt,
+                     memory::dims dst_dims, int num_inputs, int concat_dims,
+                     memory::format mkl_common_format)
+      : dst_dims(dst_dims),
+        num_inputs(num_inputs),
+        concat_dims(concat_dims),
+        mkl_common_format(mkl_common_format) {
+    for (int k = 0; k < num_inputs; k++) {
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 7]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+      std::shared_ptr<mkldnn::memory::primitive_desc> src_mpd(
+          new memory::primitive_desc(*source_md, cpu_engine_));
+      context_.src_pd_shdptr.push_back(src_mpd);
+
+      std::shared_ptr<mkldnn::memory> src_mem(
+          new mkldnn::memory(*src_mpd, DummyData));
+      context_.data_mem_shdptr.push_back(src_mem);
+    }
+
+    for (size_t i = 0; i < concat_fwd_dims.num_inputs; i++) {
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: i hope we can. will try and update

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 8]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+// needs to drop "T".
+template <typename T>
+class MklConcatFwdPrimitive : public MklPrimitive {
+ public:
+  explicit MklConcatFwdPrimitive(const MklConcatFwdParams& concat_fwd_dims,
+                                 const std::vector<memory::desc>& srcs_md)
+      : cpu_engine_(engine::cpu, 0) {
+    context_.fwd_stream.reset(new stream(stream::kind::eager));
+    // create concat primitive
+    if (context_.concat_fwd == nullptr) {
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Hope this check is not needed anyway i will check and confirm

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 9]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+
+  ~MklConcatFwdPrimitive() {}
+
+  // concat forward execute
+  //   src_data:    input data buffer of src
+  //   dst_data:    output data buffer of dst
+  void Execute(const std::vector<mkldnn::memory>& in_data,
+               const mkldnn::memory& dst_data,
+               const MklConcatFwdParams& concat_fwd_dims) {
+    CHECK_EQ(in_data.size(), context_.data_mem.size());
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 10]
```diff
@@ -235,6 +233,216 @@ class EigenConcatBaseOp : public OpKernel {
...
+    // after exec, set data handle back
+    context_.dst_mem->set_data_handle(DummyData);
+    for (int k = 0; k < concat_fwd_dims.num_inputs; k++) {
+      context_.data_mem_shdptr[k]->set_data_handle(DummyData);
+    }
+
+    for (size_t i = 0; i < concat_fwd_dims.num_inputs; i++) {
+      context_.data_mem[i] = *context_.data_mem_shdptr[i];
+    }
+    return;
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 11]
```diff
@@ -390,6 +598,11 @@ class MklConcatOp : public OpKernel {
       bool isMklReorderNeeded = false;
       memory::format mkl_common_format = memory::format::any;
       std::vector<primitive::at> inputs;
+      std::vector<memory::dims> src_dims_pt;
+
```
[the remark this answers is not in the document]
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

File: tensorflow/core/kernels/mkl_concat_op.cc (hunk 1 above)
Comment (pallavigopal) [reply tensorflow/core/kernels/mkl_concat_op.cc]: ok

**Comment by pallavigopal** [comment]
@penpornk  Thank you for your valuable review comments. I will address the review comments and submit the patch ASAP

**Comment by pallavigopal** [comment]
@penpornk Sorry for the delay. Addressed all the review comments, please review it

**Review by penpornk — APPROVED** [review]
Thank you for the fixes!

**Pull request merged**